pkg/kubelet/stats/cri_stats_provider.go

```diff
@@ -545,6 +545,9 @@ func (p *criStatsProvider) getContainerUsageNanoCores(stats *runtimeapi.Containe
 	}
 
 	nanoSeconds := stats.Cpu.Timestamp - cached.Timestamp
+	if nanoSeconds == 0 {
```

Comment:
how would this ever occur?  Does this mean two calls to the CRI stats gave metrics with the same timestamp?

Reply:
^ cc @adelina-t

Reply:
I've seen this on Windows nodes too, but not sure yet how this could've happened.
dockershim adds the timestamp itself for cpu usage by calling `time.Now()`, so it should not have the same timestamp for two calls.

Reply:
I agree with @dashpole that this is just papering over the actual bug. Given that, I'd suggest:
 - Log an error message including the content of the new stats and the cache stats for further debugging
 - return `nil` in this case to surface the problem (but not crashing kubelet of course).

Reply:
@ddebroy  any ideas on why this could have happened?

Reply:
> I agree with @dashpole that this is just papering over the actual bug. Given that, I'd suggest:
> 
> * Log an error message including the content of the new stats and the cache stats for further debugging
> * return `nil` in this case to surface the problem (but not crashing kubelet of course).

Yu-Ju, i asked for the exact same thing at the same time. we need to log the values and see what they look like

Reply:
+1 for the proposal to log the timestamps. The values are populated with time.Now().UnixNano() at https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/dockershim/docker_stats_windows.go#L89 . That is a core golang API. 

The golang docs specify using `sub()` to measure duration: https://golang.org/pkg/time/ because that uses monotonic time. This https://go.googlesource.com/proposal/+/master/design/12914-monotonic.md describes the proposal to shift to using monotonic times from wall clock and specifically mentions the usage of `sub()`. VMs may be prone to wall clock drifts. So maybe it's safer to use `sub()` and then converting the duration to nanoseconds?

Reply:
go timers are lower resolution on Windows.
(not to mention i've seen wonky behavior in the past).

if the existing process relies on nano precision, it has to be adapted for Windows i.e. possibly consider introducing small delays if needed.

> +1 for the proposal to log the timestamps

+1 here as well.

to illustrate what i mean above try comparing this between Windows and Linux:

<details>
```
package main

import (
	"fmt"
	"time"
	"math/rand"
)

func load() {
	for i := 0; i < rand.Intn(10); i++ {
		fmt.Println("i=", i)		
	}	
}

func main() {
	for i := 0; i < 10; i++ {
		t1 := time.Now()
		load()		
		t2 := time.Now()
		fmt.Println("-----------> diff", (t2.Sub(t1)).Nanoseconds())
	}
}
```
</details>

Reply:
yeah - I was going down the same path - if this isn't using time.sub() is that the problem?

I tried to track down the accuracy of time.Now()'s monotonic clock on Windows but ended up at a dead end with [runtimeNano](https://github.com/golang/go/blob/3813edf26edb78620632dc9c7d66096e5b2b5019/src/internal/poll/fd_poll_runtime.go#L19). If anyone has a pointer to the C or ASM code for this, please share. Windows does have both monotonic and wall time down to microsecond [link](https://docs.microsoft.com/en-us/windows/desktop/SysInfo/acquiring-high-resolution-time-stamps)

Reply:
@PatrickLang 
https://github.com/golang/go/blob/master/src/runtime/sys_windows_amd64.s#L476-L494

https://github.com/golang/go/blob/84066f1b0b68761a75ce9064f4c412f751ee2168/src/runtime/os_windows.go#L344-L351

Reply:
i think it might be worth a shot to force an explicit proc-ed `QueryPerformanceCounter` and don't rely on `Now()` or `Sub()`, to see if the behavior changes:
https://github.com/ScaleFT/monotime/blob/master/clock_windows_qpf.go